File: demo/design-pattern/src/design/singleObject/SingleObject.js
export class SingleObject {
    login() {
        console.log('login...')
    }
}

// 静态方法
SingleObject.getInstance = (function() {
    let instance
    return function() {
        if(!instance) {
            instance = new SingleObject()
        }
        return instance
    }
})()


// -------测试---------
// 这里只能使用静态函数 getInstance，不能 new SingleObject()
// let obj1 = SingleObject.getInstance()
// obj1.login()

// let obj2 = SingleObject.getInstance()
// obj2.login()
// console.log(obj1 === obj2) // 两者必须完全相等

File: demo/design-pattern/src/design/singleObject/SingleObject.java
public class SingleObject{
    // 注意，私有化构造函数，外部不能new，只能内部new
    private SingleObject() {}
    // 唯一被 new 出来的对象
    private SingleObject instance = null;
    // 获取对象的唯一接口
    public SingleObject getInstance() {
        if(instance == null){
            // 只 new 一次
            instance = new SingleObject();
        }
        return instance;
    }

    // 对象方法
    public void login(username, password) {
        System.out.println("login...")
    }
}